plugins/slack_app/app/controllers/slack_app_controller.rb

```
@@ -0,0 +1,158 @@
+    #   ],
+    #   "callback_id"=>"123456", <-- maps to a deployment
+    #   "user"=>{"id"=>"U0HAGH3AB", "name"=>"ben"}, <-- who clicked
+    #   "action_ts"=>"1468259216.482422",
+    #   "message_ts"=>"1468259019.000006", <-- need this to maybe update the message
+    #   "response_url"=>"https://hooks.slack.com/actions/T0HAGP0J2/58691408951/iHdjN2zjX0dZ5rXyl84Otql7"
+    #    ^^^^ This is how to respond later, we can also directly return some JSON
+    # }
+
+    # This means someone clicked the "Approve" button in the channel
+    deploy = Deploy.find_by_id @payload['callback_id']
+    return '_(Unable to locate this deploy.)_' unless deploy
+
+    # Is this user connected?
+    return {
+      replace_original: false,
+      text: unknown_user
+    } unless @current_user_from_slack
+
+    # Can this user approve the deployment?
+    return {
+      replace_original: false,
+      text: "You cannot approve your own deploys."
+    } if @current_user_from_slack == deploy.user
+    return {
+      replace_original: false,
+      text: "You do not have permissions to approve that deploy."
+    } unless @current_user_from_slack.deployer_for?(deploy)
+
+    # Buddy up
+    deploy.confirm_buddy!(@current_user_from_slack)
+    SlackMessage.new(deploy).message_body
+  end
+
+  def unknown_user
+    "You're not set up to use this yet. Please visit " \
+    "#{url_for action: :oauth, only_path: false} " \
+    "to connect your accounts."
+  end
+
+  def unknown_project(project)
+    "Couldn't find a project with permalink `#{project}`."
+  end
+
+  def unknown_stage(project, stage_name)
+    "Sorry, " \
+    "<#{project_path(project)}|#{project.name}> " \
+    "doesn't have a stage named `#{stage_name}`."
+  end
+
+  def unauthorized_deployer(project)
+    "Sorry, it doesn't look like you have permission to create a deployment on `#{project}`."
+  end
+
+  def verify_environment
+    raise "Invalid configuration" unless ENV['SLACK_VERIFICATION_TOKEN'].present?
```

maybe `raise "SLACK_VERIFICATION_TOKEN is not set"` to make it more obvious what needs fixing